catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/MetacardFactory.java

```diff
@@ -41,66 +46,80 @@
 
-  public MetacardFactory(
-      MimeTypeToTransformerMapper mimeTypeToTransformerMapper, UuidGenerator uuidGenerator) {
-    this.mimeTypeToTransformerMapper = mimeTypeToTransformerMapper;
+  public MetacardFactory(UuidGenerator uuidGenerator, TransformerLocator transformerLocator) {
     this.uuidGenerator = uuidGenerator;
+    this.transformerLocator = transformerLocator;
   }
 
-  Metacard generateMetacard(String mimeTypeRaw, String id, String fileName, Path tmpContentPath)
+  List<Metacard> generateMetacard(
+      String mimeTypeRaw,
+      String id,
+      String fileName,
+      Path tmpContentPath,
+      Map<String, ? extends Serializable> arguments)
       throws MetacardCreationException, MimeTypeParseException {
 
-    Metacard generatedMetacard = null;
-
-    MimeType mimeType = new MimeType(mimeTypeRaw);
-
-    List<InputTransformer> listOfCandidates =
-        mimeTypeToTransformerMapper.findMatches(InputTransformer.class, mimeType);
     List<String> stackTraceList = new ArrayList<>();
 
-    LOGGER.debug("List of matches for mimeType [{}]: {}", mimeType, listOfCandidates);
+    MimeType mimeType = new MimeType(mimeTypeRaw);
 
-    for (InputTransformer candidate : listOfCandidates) {
-      try (InputStream transformerStream =
-          com.google.common.io.Files.asByteSource(tmpContentPath.toFile()).openStream()) {
-        generatedMetacard = candidate.transform(transformerStream);
-      } catch (CatalogTransformerException | IOException e) {
-        List<String> stackTraces = Arrays.asList(ExceptionUtils.getRootCauseStackTrace(e));
-        stackTraceList.add(String.format("Transformer [%s] could not create metacard.", candidate));
-        stackTraceList.addAll(stackTraces);
-        LOGGER.debug("Transformer [{}] could not create metacard.", candidate, e);
-      }
-      if (generatedMetacard != null) {
-        break;
-      }
-    }
+    Optional<List<Metacard>> generatedMetacards =
+        transformerLocator
+            .findMultiInputTransformers(mimeType)
+            .stream()
+            .map(
+                multiInputTransformer ->
+                    attemptTransform(
+                        multiInputTransformer, tmpContentPath, stackTraceList, arguments))
+            .filter(Objects::nonNull)
+            .findFirst();
 
-    if (generatedMetacard == null) {
+    if (!generatedMetacards.isPresent()) {
       throw new MetacardCreationException(
           String.format(
               "Could not create metacard with mimeType %s : %s",
               mimeTypeRaw, StringUtils.join(stackTraceList, "\n")));
     }
 
-    if (id != null) {
-      generatedMetacard.setAttribute(new AttributeImpl(Metacard.ID, id));
-    } else {
-      generatedMetacard.setAttribute(new AttributeImpl(Metacard.ID, uuidGenerator.generateUuid()));
-    }
+    return generatedMetacards
+        .get()
+        .stream()
+        .peek(
+            metacard -> {
+              if (id != null) {
```

Comment:
✏️ My personal preference is to extract large blocks like this into private methods so they don't introduce noise to the streaming logic itself. It's not a requirement, but I find it reads more easily.